spec/postgres_runner_helper.rb

```diff
@@ -0,0 +1,142 @@
+require "etc"
+require "tmpdir"
+require "socket"
+require "fileutils"
+require "singleton"
+
+require_relative 'support/mixins/pg_environment_updater'
+
+# This class is in charge of booting additional instance of postgres that can
+# be used for validating features of PostgresAdmin.
+#
+# It makes use of `pg_ctl` to init, start and stop the instance, in is built so
+# that it can be used as a stub for LinuxAdmin as well.  While this is a
+# singleton, this really should only booted at the beginning of the suite, and
+# shutdown and cleaned up prior to exiting.
+class PostgresRunner
+  include Singleton
+  include FileUtils
+  include PgEnvironmentUpdater
+
+  attr_accessor :db_dir, :port
+
+  def self.start
+    instance.start
+  end
+
+  def self.stop
+    instance.stop
+  end
+
+  def self.running?
+    instance.running?
+  end
+
+  def self.port
+    instance.port
+  end
+
+  def self.set_env
+    instance.set_env
+  end
+
+  def self.destroy
+    instance.stop
+    instance.clean
+  end
+
+  def self.user
+    Etc.getlogin
+  end
+
+  def self.group
+    Etc.getgrgid(Etc.getpwnam(Etc.getlogin).gid).name
+  end
+
+  def initialize
+    initdb
+    set_env
+    set_port
+    start
+    create_root_role
+    create_postgres_role
+  end
+
+  def start
+    return if running?
+    fix_pg_conf
+    system("#{pg_ctl} start -D #{db_dir} -wo '-p #{port}'", :out => File::NULL)
+  end
+
+  def stop
+    return unless running?
+    system("#{pg_ctl} stop -D #{db_dir} -wm fast", :out => File::NULL)
+  end
+
+  def clean
+    remove_entry(db_dir)
+  end
+
+  def running?
+    return false unless db_dir
+    # avoiding a shellout by just checking the existance of the pg_ctl pid file
+    #
+    # see `man pg_ctl` in the FILES section for the description of this file
+    File.exist?(pid_file)
+  end
+
+  def set_env
+    @user = self.class.user # used by PgEnvironmentUpdater
+    ENV["APPLIANCE_PG_DATA"]    = db_dir
+    ENV["APPLIANCE_PG_SERVICE"] = "local_pg_instance"
+  end
+
+  private
+
+  def set_port
+    @port = TCPServer.open("127.0.0.1", 0) { |sock| sock.addr[1] }
+  end
+
+  def pid_file
+    File.join(db_dir, "postmaster.pid")
+  end
+
+  def initdb
+    dirname = Dir::Tmpname.make_tmpname("postgres_admin", "data")
+    @db_dir = File.expand_path(dirname, __dir__)
+
+    system("#{pg_ctl} init -D #{db_dir} -o '-A trust'", :out => File::NULL)
```

Comment:
Would specifying `-U` to `initdb` here simplify some things around setting up users? If we always initialize the database with a consistent superuser then maybe we don't have to work around how things will work differently locally vs on travis.

Reply:
Well, again, on Travis (currently), we are trying to use the existing database instance, where locally we are trying to not blow away the developers data when doing a base backup restore.  If we wanted to keep this the same, basically both Travis and local would have to use the `pg_ctl` method, which is actually probably fine.  I tried not going that route originally because I did want to try working with an existing database when doing a restore when possible.

That said, this would probably avoid all of the `sudo ruby -r postgres_admin` stuff I do, so this might not be a bad idea.  The `-U` would be unnecessary, though, as we would just use the user that is running the specs be the owner of the DB and leave it at that.